forester/src/epoch_manager.rs

```diff
@@ -272,24 +257,61 @@ impl<R: RpcConnection, I: Indexer<R>> EpochManager<R, I> {
     async fn process_epoch(&self, epoch: u64) -> Result<()> {
         info!("Entering process_epoch");
 
-        // Registration
-        let mut registration_info = self.register_for_epoch(epoch).await?;
+        let processing_flag = self
+            .processing_epochs
+            .entry(epoch)
+            .or_insert_with(|| Arc::new(AtomicBool::new(false)));
+
+        if processing_flag
+            .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
+            .is_err()
+        {
+            // Another task is already processing this epoch
+            debug!("Epoch {} is already being processed, skipping", epoch);
+            return Ok(());
+        }
+
+        // Attempt to recover registration info
+        let mut registration_info = match self.recover_registration_info(epoch).await {
```

Comment:
what does this do if there is no registration that can be recovered?

Reply:
If there is no registration we'll try to register here https://github.com/Lightprotocol/light-protocol/pull/1160/files/7213266f76690b8931a73e01055af9123b07b22a#diff-1e165bd3189acab768bdb1cd6cf4ec33528400511ec68498004b73a25d0646c3R280-R281